SalesforceCommerceCloud/pwa-kit
# W-15040197 Lazy basket initialization with feature flag

# Description

Current: today the PWA Kit calls SCAPI endpoints to create a basket on session start. This has two consequences:

- An extra API that may/may not be helpful, this takes up more network bandwidth. A user may not interact with the baskets until they add an item to basket
- In hybrid sites create the possibility of a condition between cart instantiation and retrieval between SFRA and PWA Kit

This PR added a helper function in commerce-sdk-react `useShopperBasketsMutationHelper` that returns a map of helpers. Currently, it has `addItemToNewOrExistingBasket` that is responsible for managing the process of adding an item to a basket.
- If a basket already exists, add the item to the basket immediately.
 - If a basket does not exist, create a new basket using the createBasket mutation
and then add the item to the newly created basket using the addItemToBasket mutation.

`addAnItemToCart` is the only mutation that needs to be aware of basket creation. The other mutations (e.g removeItem or updateItem or the rest) are actions that always come after `addAnItemToCart` has been called. 

To avoid breaking change, the behavior is behind a feature flag `LAZY_BASKET_INITIALIZATION` that allows current behavior to stay the same. Users can turn of lazy basket initialization by flipping the flag



# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- check out code
- Start retail app
- Clear application storage
- By default `LAZY_BASKET_INITIALIZATION` is false. Confirm that the change does not break current implementation

// Test lazy basket initialization
- change value of LAZY_BASKET_INITIALIZATION to true
- Clear app storage
- Refresh the page
- Confirm that basket is not created until an item is not created on PDP and wishlist page

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
